### DISCUSSION TITLE
JS Client should use fetch

### DISCUSSION THREAD

**Discussion by alightinastorm** [medusajs/medusa#4749] [Feature Requests]
not much to say, it should use fetch instead of axios (which uses XMLHttpRequest internally)

Nextjs 13 App directory patches native fetch for caching, which is completely lost when using XMLHttpRequest

And the client should be extensible, eg:

```
const client = new Medusa({})

client.admin.extend({ custom: { 
   retrieve: async (params) => { 
 // custom logic
  },
})

```

This way actually extending the medusa backend can be used with the existing client, instead of having two separate clients

I'd volunteer going down that rabbit hole

**Comment by srindom** [comment]
The client being axios-powered is a pet-peeve of mine, but we haven't gotten to revisit it yet.

The extension idea is excellent too. Would love to see a PR on this - happy to help with getting started if interested.

**Reply by alightinastorm** [reply]
To be honest, changing the current client is nearly impossible to remove axios, since on every entity there is a "request" object which can be used to call custom queries

Removing axios might break many existing apps relying on that

Also, a general question is how could we include the types of the api of custom entities when extending?

Currently we are manually copying them, but that is unmaintainable in the long run as they get out of sync on every change

I saw that medusa is preparing to switch to mikroorm, are there any plans to abstract the definition of models and repositories to a medusa-owned layer and make the actual orm an adapter to that?

Sorry for exploding this topic, just trying to find the right angle to tackle this, since simply switching fetch is a rather easy task but the underlying problems might be bigger (as in, developer experience problems)

Edit: Just found out about the plans of having a data access layer
